### PR TITLE
fix: use correct encoding for L1 data fee of EIP2718 transactions

### PR DESCRIPTION
**Pull request by Thegaram**
## 1. Purpose or design rationale of this PR

[EIP2718](https://eips.ethereum.org/EIPS/eip-2718) defines the byte encoding as `Type || Payload`, but `EncodeRLP ` returns an actual RLP array, i.e. it prepends the marshaled result with one or more bytes. We should use `MarshalBinary` instead. Note: The behavior of `EncodeRLP ` and `MarshalBinary` is identical for legacy transactions, so this is not a breaking change.

This PR also adds an upper bound on `L1DataFee` to ensure these values don't overflow in the tracing/circuit code. Technically, this is a breaking change, but in practice such high data fee values never occur on Scroll.

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [X] fix: A bug fix

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [X] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [X] This PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: rollup/fees/rollup_fee.go [hunk 1]
```diff
@@ -40,7 +41,6 @@ type Message interface {
 // required to compute the L1 fee
 type StateDB interface {
 	GetState(common.Address, common.Hash) common.Hash
-	GetBalance(addr common.Address) *big.Int
```
Comment (omerfirmak) [inline rollup/fees/rollup_fee.go:43]: nit: looks unrelated?

**Review by 0xmountaintop — DISMISSED** [review]

File: rollup/fees/rollup_fee.go [hunk 2]
```diff
@@ -216,7 +206,7 @@ func CalculateL1DataFee(tx *types.Transaction, state StateDB, config *params.Cha
 		return big.NewInt(0), nil
 	}
 
-	raw, err := rlpEncode(tx)
+	raw, err := tx.MarshalBinary()
```
Comment (omerfirmak) [inline rollup/fees/rollup_fee.go:209]: `MarshalBinary`still encodes the type information, right?

File: rollup/fees/rollup_fee.go [hunk 3]
```diff
@@ -231,6 +221,12 @@ func CalculateL1DataFee(tx *types.Transaction, state StateDB, config *params.Cha
 		l1DataFee = calculateEncodedL1DataFeeCurie(raw, gpoState.l1BaseFee, gpoState.l1BlobBaseFee, gpoState.commitScalar, gpoState.blobScalar)
 	}
 
+	// ensure l1DataFee fits into uint64 for circuit compatibility
+	// (note: in practice this value should never be this big)
+	if !l1DataFee.IsUint64() {
+		l1DataFee = new(big.Int).SetUint64(math.MaxUint64)
```
Comment (omerfirmak) [inline rollup/fees/rollup_fee.go]: nit:

```suggestion
		l1DataFee.SetUint64(math.MaxUint64)
```

File: rollup/fees/rollup_fee.go (hunk 1 above)
Comment (Thegaram) [reply rollup/fees/rollup_fee.go:43]: Correct, it's unrelated. Originally I planned to add some L1 data fee unit tests with a mock state, that's when I noticed that this part of the interface is not used. We might still add those tests later, but for this PR I decided to add tests in `transaction_tests` instead.

File: rollup/fees/rollup_fee.go (hunk 2 above)
Comment (Thegaram) [reply rollup/fees/rollup_fee.go:209]: Yes. Both `tx.MarshakBinary` and `tx.EncodeRLP` use `tx.encodeTyped` internally, which prepends the type byte. The difference is that the former directly returns `buf.Bytes()`, while the latter returns it RLP-encoded (`rlp.Encode(w, buf.Bytes())`).

**Review by omerfirmak — APPROVED** [review]
